sql/core/src/main/scala/org/apache/spark/sql/internal/SharedState.scala

```diff
@@ -191,11 +195,18 @@ private[sql] class SharedState(
 }
 
 object SharedState extends Logging {
-  try {
-    URL.setURLStreamHandlerFactory(new FsUrlStreamHandlerFactory())
-  } catch {
-    case e: Error =>
-      logWarning("URL.setURLStreamHandlerFactory failed to set FsUrlStreamHandlerFactory")
+  private val fsUrlStreamHandlerFactoryInitialized = new AtomicBoolean(false)
+
+  private def setFsUrlStreamHandlerFactory(conf: SparkConf): Unit = {
+    if (conf.get(DEFAULT_URL_STREAM_HANDLER_FACTORY_ENABLED) &&
+        fsUrlStreamHandlerFactoryInitialized.compareAndSet(false, true)) {
+      try {
+        URL.setURLStreamHandlerFactory(new FsUrlStreamHandlerFactory())
```

Comment:
With sync, do you think it's possible that the other thread comes in seeing `fsUrlStreamHandlerFactoryInitialized == true` but the actual initialization is not finished?

Reply:
This is AtomicBoolean, @dbtsai ~

Reply:
Highly unlikely but possible. It's worth a simple defensive check against a flag.

Reply:
`AtomicBoolean` only guarantees that the following `try catch` block will only be executed once in the current thread, but it doesn't guarantee that the other thread constructs a new session without waiting the current initialization finishes. 

As @srowen says, it's unlikely to happen, but possible logically. Will be great to prevent it. I feel the an implementation using `volatile boolean` and `sync` with double checked pattern will be ideal for this case.

Reply:
Or the other way we can possibly do by using `AtomicBoolean` is having the `try catch` block moved into 
```scala
fsUrlStreamHandlerFactoryInitialized.getAndSet {
  ...
  true
}
```
so it will be guarded.

Reply:
I'm also more familiar with the double-check pattern. As we see, the original code (before AtomicBoolean) was double-check pattern with `volatile` and `synchronized`. I'm fine to revert to that. If we use double-checked pattern, `AtomicBoolean` is useless technically.

BTW, `fsUrlStreamHandlerFactoryInitialized` is inside `SharedState` object . What do you mean by `new session`? The new thread also need to pass `getAndSet` on the same `fsUrlStreamHandlerFactoryInitialized`.

Reply:
We need to set `isInitialized = true` after the `URL.setURLStreamHandlerFactory`, so I feel double-check pattern is much more preferable than `AtomicBoolean`.

I mean the other thread constructing `class SharedState`. 

Yeah, let's say two threads are constructing `class SharedState`, and the first thread set `isInitialized = true`, and then starts to run actual `URL.setURLStreamHandlerFactory`. The second thread see `isInitialized == true`, and finishes the construction before the first thread actually initializes `URL.setURLStreamHandlerFactory`, this will cause problem.

In double check pattern, this is not concern since the whole thing will be locked until the first thread finishes the initialization.

Reply:
We can do `volatile boolean` here since we dont need to keep the reference.

Reply:
I see your point. In that case a simple volatile boolean is fine for the flag.